File: solutions/prospective/algorithms/CLRS/js/matrix_multiply_naive.js
/**
 * Naive Matrix Multiplication
 * 
 * Time Complexity: O(n^3) for n x n matrices
 * Space Complexity: O(n^2) for result
 * 
 * Multiplies two matrices A and B.
 */
// === MEMO START ===
function matrixMultiply(A, B) {
    const rowsA = A.length;
    const colsA = A[0].length;
    const rowsB = B.length;
    const colsB = B[0].length;

    if (colsA !== rowsB) {
        throw new Error("Incompatible dimensions");
    }

    const C = Array(rowsA).fill().map(() => Array(colsB).fill(0));

    for (let i = 0; i < rowsA; i++) {
        for (let j = 0; j < colsB; j++) {
            for (let k = 0; k < colsA; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    return C;
}

File: solutions/prospective/algorithms/CLRS/c/matrix_multiply_naive.c
/*
Naive Matrix Multiplication

Time Complexity: O(n^3) for n x n matrices
Space Complexity: O(n^2) for result

Multiplies two matrices A and B.
Returns a new dynamically allocated matrix.
*/
#include <stdlib.h>

// === MEMO START ===
int** matrix_multiply(int **A, int rows_A, int cols_A, int **B, int rows_B, int cols_B) {
    if (cols_A != rows_B) {
        return NULL; // Error: Incompatible dimensions
    }

    int **C = (int **)malloc(rows_A * sizeof(int *));
    for (int i = 0; i < rows_A; i++) {
        C[i] = (int *)calloc(cols_B, sizeof(int));
    }

    for (int i = 0; i < rows_A; i++) {
        for (int j = 0; j < cols_B; j++) {
            for (int k = 0; k < cols_A; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    return C;
}

